tools/audio/sampleconv/src/codec/vadpcm.c

```diff
@@ -0,0 +1,1324 @@
+
+            vencodeframe(&outdata[nBytes], in_buf, state, coef_tbl, order, npredictors, frame_size);
+            nBytes += frame_size;
+        }
+
+        // Emplace loop state
+        // printf("    wr loop state\n");
+        for (j = 0; j < 16; j++) {
+            if (state[j] > 0x7FFF)
+                state[j] = 0x7FFF;
+            if (state[j] < -0x7FFF)
+                state[j] = -0x7FFF;
+            aloops[i].state[j] = state[j];
+        }
+
+        aloops[i].count = -1;
+
+        // Encode the loop for n repeats
+        while (nRepeats > 0) {
+            // printf("    repeat #%u\n", nRepeats);
+            // Encode [loop_start, loop_end]
+            while (currentPos + 16 < aloops[i].end) {
+                // printf("    loop-repeat %u / %u\n", currentPos, aloops[i].end);
+
+                memcpy(in_buf, &indata[currentPos], sizeof(in_buf));
+                currentPos += 16;
+
+                vencodeframe(&outdata[nBytes], in_buf, state, coef_tbl, order, npredictors, frame_size);
+                nBytes += frame_size;
+            }
+
+            // Handling for when loop_end is halfway through a frame
+
+            uint32_t left = aloops[i].end - currentPos;
+            // printf("    rem=%u\n", left);
+
+            memcpy(in_buf, &indata[currentPos], left * sizeof(int16_t));
+
+            currentPos = aloops[i].start * 2;
+
+            memcpy(in_buf + left, &indata[currentPos], (16 - left) * sizeof(int16_t));
+
+            vencodeframe(&outdata[nBytes], in_buf, state, coef_tbl, order, npredictors, frame_size);
+            nBytes += frame_size;
+
+            // Return to loop start
+
+            currentPos = aloops[i].start - left + 16;
+            nRepeats--;
+        }
+        aloops[i].end = newEnd;
+    }
+
+    // Get frame count, truncate to loop end if desired
+    if (nloops > 0 && opts->truncate)
+        nFrames = MIN(aloops[nloops - 1].end + 16, nFrames);
+
+    // Encode remaining data
+
+    while (currentPos < nFrames) {
+        uint32_t nsam = MIN(nFrames - currentPos, 16);
+        // printf("    post-loop: %u / %u (%u)\n", currentPos, nFrames, nsam);
+
+        memcpy(in_buf, &indata[currentPos], sizeof(int16_t) * nsam);
+        currentPos += nsam;
+
+        /* printf("Padding with %u 0s\n", 16 - nsam); */
+        memset(in_buf, 0, (16 - nsam) * sizeof(int16_t));
+
+        vencodeframe(&outdata[nBytes], in_buf, state, coef_tbl, order, npredictors, frame_size);
+        nBytes += frame_size;
+    }
+
+    // Pad to even number
+    if (nBytes % 2 != 0)
+        outdata[nBytes++] = 0;
+
+    assert(nBytes <= ctnr->data_size);
+
+    // Write out
+
+    ctnr->num_loops = 0;
+    if (ctnr->loops != NULL)
+        free(ctnr->loops);
+    ctnr->loops = NULL;
+
+    free(ctnr->data);
+
+    ctnr->data = outdata;
+    ctnr->data_size = nBytes;
+    ctnr->data_type = (frame_size == 5) ? SAMPLE_TYPE_VADPCM_HALF : SAMPLE_TYPE_VADPCM;
+    // ctnr->num_samples = (nFrames + 15) & ~15;
```

Comment:
What would this commented assignment be for?

Reply:
It would update `num_samples` to cover every frame that was consumed during encoding, but I vaguely recall having some issues with updating it in this way so I left it at the number of samples that the original input file reported.

Reply:
Thanks. Personally I would recommend just removing it then (though you know my stance on dead cod :) ). Will leave it up to you.

Reply:
Done